third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/TensorMemoryToLLVM.cpp

```diff
@@ -12,9 +12,62 @@ using namespace mlir;
 using namespace mlir::triton;
 using namespace mlir::triton::gpu;
 
-static const int largestTmemLoadStore = 128;
+// The maximum number of registers that can be accessed by a single message
+// regardless of shape or repetitions
+static constexpr int largestTmemLoadStore = 128;
+// Factor used to intentionally narrow contiguous tmem access down from
+// the maximum message width to avoid register pressure
+static constexpr int narrowingFactor = 4;
+
 namespace {
 
+// Tensor memory access traits, currently only 32b is used
+template <bool IsStrided> struct TMemAccess32b {
+  static constexpr int opBitWidth = 32;
+  static constexpr int colsPerThread = 1;
+  static constexpr int rowsPerThread = 1;
+  static constexpr const char *opShape = IsStrided ? "16x32bx2" : "32x32b";
+  static constexpr bool usesSecondHalfOffset = IsStrided;
+};
+
+struct TMemAccess256b {
+  static constexpr int opBitWidth = 256;
+  static constexpr int colsPerThread = 2;
+  static constexpr int rowsPerThread = 2;
+  static constexpr char opShape[] = "16x256b";
+};
+
+template <typename TMemAccess> struct TMemMessage {
+  using traits = TMemAccess;
+  static constexpr int opBitWidth = TMemAccess::opBitWidth;
+  static constexpr int colsPerThread = TMemAccess::colsPerThread;
+  static constexpr int rowsPerThread = TMemAccess::rowsPerThread;
+  static constexpr int numThreadsPerWarp = 32;
+  static constexpr int maxNumRepeats =
+      largestTmemLoadStore / (colsPerThread * rowsPerThread);
+  static constexpr int maxColsPerMessage = (opBitWidth / 32) * maxNumRepeats;
+  static constexpr int rowsPerMessage = numThreadsPerWarp / rowsPerThread;
+  static constexpr int colsPerMessage = maxColsPerMessage / narrowingFactor;
+};
```

Comment:
The templates seems a bit of an overkill, any chance we can remove them?

Reply:
Sure, my only goal was to separate (1) tmem atom derived message constants from (2) workload derived message constraints. I chose compile time constant expressions to make the separation, but have updated to use runtime params. Let me know if this is an acceptable option to you that also meets my goal.

Reply:
fair enough, maybe that's me disliking templates too much ;) thanks for changing it